### DISCUSSION TITLE
Recovering computational parameters with custom network

### DISCUSSION THREAD

**Discussion by Harris-D** [ilabcode/pyhgf#218] [Q&A] [answered]
Hi, I have made a custom network structure and want to fit it to observed data from an experiment to recover estimates of model parameters. I can't work out if this is possible for custom networks, or if only for standard 2 and 3 level HGFs? I.e., i have the known inputs plus observed behaviours. Thanks

**Comment by LegrandNico** [comment, marked as answer]
Hi @Harris-D 

Thank you for reaching out. It is indeed possible to sample any custom network to infer parameter estimates, it simply requires:
- writing a response function
- wrapping this response function in a PyTensor Op so it can be inserted in a PyMC graph.

You can find details on how to write a response function in [this tutorial](https://ilabcode.github.io/pyhgf/notebooks/2-Using_custom_response_functions.html) - the response function should return the log probability of your observation given a set of parameters.

Then you might create a PyMC Op for this function, and the gradient of this function. You can find an example in the [categorical HGF tutorial](https://ilabcode.github.io/pyhgf/notebooks/1.2-Categorical_HGF.html) and in [this notebook](https://github.com/ilabcode/pyhgf/blob/c48709a77aed39e063c9ec229ff2493c444bd5d0/docs/source/notebooks/Example_4_Iowa_Gambling_Task_Short.ipynb)  (using the grad function, that is easier to work with). You can also refer to these tutorials from the PyMC team: [here](https://www.pymc-labs.com/blog-posts/jax-functions-in-pymc-3-quick-examples/) and [here](https://www.pymc.io/projects/examples/en/latest/howto/wrapping_jax_function.html).

Let me know if you manage to make this work.

Cheers,

**Reply by Harris-D** [reply]
Hi Nicolas, thank you for your reply. It took me a while but I have got it working, thank you!
